Corner plot - close #26

mkborregaard opened the pull request:
This works now in the sense that
```julia
using Turing, StatPlots, MCMCChain
@model f(x) = begin
          A ~ Normal(0,1)
          B ~ Normal(0,1)
          C ~ Normal(0,1)
          x ~ Normal(0,1)
       end
chain = sample(f(1), HMC(500,0.01,1))
chaincorner(chain, [:A, :B, :C])
```
gives
<img width="580" alt="skaermbillede 2019-01-17 kl 21 29 15" src="https://user-images.githubusercontent.com/8429802/51347236-d3f9da80-1a9f-11e9-9943-1dff9c4787c1.png">

But, I really think we should align the syntax with the rest of the plotting syntax you have here. I can't 100% see the design of that so would be good with input. @trappmartin @cpfiffer 

Also, this requires a change in StatPlots so will require StatPlots 0.9, and will only work on julia-1.0+ 
That is a common issue with advanced recipes.

Finally, shouldn't the default size of those points in cornerplot not be smaller? :-O

mkborregaard commented:
Changed the syntax to simply
```
plot(mychain, seriestype = :corner)
```
will currently not work so well because `keys` gives too many names, but that should be fixed in `keys` (I think).
Requires the most recent release of StatPlots (0.9)

cpfiffer commented:
Does the way these recipes are specified allow for us to add a keyword argument to `plot`? Can we support something like `plot(chain, seriestype = :corner, params = [:A, :B, :C])`?

Also, thank you for your help, @mkborregaard! It would take me what I suspect is a very long time to assemble this.

mkborregaard commented:
Yes it does allow that - it's only a question of choosing the best coherent syntax. `plot(chain, [:A, :B, :C], seriestype = :corner)` would be another possibility. In both cases it would be relevant to think about whether these are useful for all the other seriestypes as well - e.g. if `colordim = :chain` would it make sense to also select parameters from the other seriestypes?
btw `parameter` is a much better keyword than `param` imho - the short lingo-forms remind me of R 🙀 :-)

cpfiffer commented:
Yes, `plot(chain, [:A, :B, :C], seriestype = :corner)` should work in all cases. If nothing is provided for the symbol vector, then it should default to using `keys(chain)` for the values to plot. That'd also make all this more future-proof. So I suppose we don't actually need a keyword argument for symbols.

mkborregaard commented:
I've updated the syntax to
```
plot(chain, parameters = [:A, :B, :C], seriestype = :corner)
# or simply
corner(chain, parameters = [:A, :B, :C])
```
It works now so can be merged if that's the final syntax. 
I think this could be improved to also allow the other plot types to accept these keyword args, to set colordim automatically to :chain when `parameters` is given, etc.

I've for starters now also added the syntax
```
plot(chain, :A, seriestype = A)
```
for `A in (traceplot, meanplot, density, histogram, autocorplot)`

Also added a code snippet to the readme, and removed the earlier verbose deprecations.

```

mkborregaard commented:
Haha sorry just saw that you'd commented you did NOT want a keyword argument for parameters. Sorry! Fixing.

mkborregaard commented:
OK, updated.
All the seriestypes now take a symbol or vector of symbols (which will be interpreted as parameter names and require colordim to be :chain), an Int (it already did) or Vector{Int}, which will be interpreted as indices into `keys(chain)`. The single-argument form is not touched.

Not sure the code is so clean anymore :-O There's a lot of checking, as corner is not compatible with the other seriestypes etc. Not really sure how to clean that up nicely.

Also note that on my computer the `corner` code **segfaults** if some keys are selected that are not the parameters. Really - not sure what's going on there.

yebai commented:
Excellent work. Ready to merge whenever you guys are happy!

mkborregaard commented:
Great. IMHO it's good to merge :shipit:
But you may want to play around with this for real models from the wild, I've only played with the toy model above.

benelsen commented on src/plot.jl:
```diff
@@ -93,47 +96,57 @@ end
+        for (j, ptype) in enumerate(ptypes)
+            for (i, par) in enumerate(parameters)
+                @series begin
+                    @show parameters
```
I believe this line still needs to go or `plot(chain)` will be a bit too verbose.

mkborregaard replied:
:joy: ouch - good catch!

mkborregaard replied:
fixed

benelsen commented:
I tested it on https://github.com/StanJulia/StatisticalRethinking.jl/blob/master/notebooks/12/m12.6.ipynb and after realising this is in fact the compact version of the StatPlots' `cornerplot`, it looks right to me 
![screenshot from 2019-01-22 22-04-19](https://user-images.githubusercontent.com/614874/51566799-df7d4500-1e95-11e9-8d55-cd19f17b9907.png)

mkborregaard commented:
Great - btw I plan to change the default colorscheme in StatPlots. I'm not sure people realize this, but the color indicates the strength of the correlation. By using a different colormap, we now have blue for positive correlations, red for negative, yellow for neutral. Comments? This is unrelated to this PR.

<img width="587" alt="skaermbillede 2019-01-22 kl 22 49 27" src="https://user-images.githubusercontent.com/8429802/51567722-32f09280-1e98-11e9-89aa-fd1fc26d3eec.png">

cpfiffer commented:
I did not realize that the colors indicated correlation! I'm fond of the new colors.
